skimage/measure/_marching_cubes_lewiner_cy.pyx

```diff
@@ -0,0 +1,1463 @@
+Journal of Graphics Tools 8(2): pp. 1-15 (december 2003)
+
+This algorithm has the advantage that it provides topologically correct
+results, and the algorithms implementation is relatively simple. Most
+of the magic is in the lookup tables, which are provided as open source.
+
+Originally implemented in C++ by Thomas Lewiner in 2002, ported to Cython
+by Almar Klein in 2012. Adapted for scikit-image in 2016.
+
+""" 
+
+# Cython specific imports
+import numpy as np
+cimport numpy as np
+import cython 
+
+# Enable low level memory management
+# from libc.stdlib cimport malloc, free
+cdef extern from "stdlib.h": # The cimport does not work on my Linux Laptop
+   void free(void* ptr)
+   void* malloc(size_t size)
+
+# Type defs, we support float32 and float64
+ctypedef np.float32_t FLOAT32_T
+ctypedef np.float64_t FLOAT64_T
+ctypedef np.int32_t INT32_T
+FLOAT32 = np.float32
+FLOAT64 = np.float64
+INT32 = np.float32
+
+
+# Define tiny winy number
+cdef double FLT_EPSILON = np.spacing(1.0) #0.0000001
+
+
+# Define abs function for doubles
+cdef inline double dabs(double a): return a if a>=0 else -a
+cdef inline int imin(int a, int b): return a if a<b else b
+
+# todo: allow dynamic isovalue?
+# todo: can we disable Cython from checking for zero division? Sometimes we know that it never happens!
+
+def remove_degenerate_faces(vertices, faces, *arrays):
+    
+    vertices_map0 = np.arange(len(vertices), dtype=np.int32)
+    vertices_map1 = vertices_map0.copy()
+    faces_ok = np.ones(len(faces), dtype=np.int32)
+    
+    cdef np.ndarray[FLOAT32_T, ndim=2] vertices_ = vertices
+    cdef np.ndarray[FLOAT32_T, ndim=1] v1, v2, v3
+    cdef np.ndarray[INT32_T, ndim=2] faces_ = faces
+    cdef np.ndarray[INT32_T, ndim=1] face_
+    
+    
+    cdef np.ndarray[INT32_T, ndim=1] vertices_map1_ = vertices_map1
+    cdef np.ndarray[INT32_T, ndim=1] faces_ok_ = faces_ok
+    
+    cdef int j, i1, i2, i3
+    
+    # Iterate over all faces. When we encounter a degenerate triangle,
+    # we update the vertex map, i.e. we merge the corresponding vertices.
+    for j in range(faces_.shape[0]):
+        face_ = faces_[j]
+        i1, i2, i3 = face_[0], face_[1], face_[2]
+        v1, v2, v3 = vertices_[face_[0]], vertices_[face_[1]], vertices_[face_[2]]
+        if v1[0] == v2[0] and v1[1] == v2[1] and v1[2] == v2[2]:
+            vertices_map1_[i1] = vertices_map1_[i1] = imin(vertices_map1_[i1], vertices_map1_[i2])
+            faces_ok_[j] = 0
+        if v1[0] == v3[0] and v1[1] == v3[1] and v1[2] == v3[2]:
+            vertices_map1_[i1] = vertices_map1_[i3] = imin(vertices_map1_[i1], vertices_map1_[i3])
+            faces_ok_[j] = 0
+        if v2[0] == v3[0] and v2[1] == v3[1] and v2[2] == v3[2]:
+            vertices_map1_[i2] = vertices_map1_[i3] = imin(vertices_map1_[i2], vertices_map1_[i3])
+            faces_ok_[j] = 0
+    
+    # Create mask and mapping to new vertex indices
+    vertices_ok = vertices_map1 == vertices_map0
```

Comment:
Instead of keeping these both in memory and then doing a binary comparison at the end to see what changed, I think you could preallocate that binary array (`vertices_ok`) off the bat and set values to True or False as desired. That should increase efficiency a bit.